File: 297_Serialize_and_Deserialize_Binary_Tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# Time Complexity : serialize - O(n) [n = number of nodes in Tree], deserialize - O(n) [n = number of elements in comma separated list] 
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
from collections import deque
class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:return []
        res = []
        q = deque([root])
        res = [str(root.val)]
        while q:
            curr = q.popleft()
            if curr.left:
                res.append(str(curr.left.val))
                q.append(curr.left)
            else:
                res.append('null')
                
            if curr.right:
                res.append(str(curr.right.val))
                q.append(curr.right)
            else:
                res.append('null')
            
        return ",".join(res)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if not data:return None
        data = data.split(',')
        root = TreeNode(data[0])
        q = deque([root])
        i = 1
        while q and i<len(data):
            curr = q.popleft()
            if i<len(data) and data[i] != 'null':
                node = TreeNode(data[i])
                curr.left = node
                q.append(curr.left)
            i+=1
            
            if i<len(data) and  data[i] != 'null':
                node = TreeNode(data[i])
                curr.right = node
                q.append(curr.right)
            i+=1
        return root
        
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: 297-SerializeAndDeserializeBinaryTree.cpp
/* 
    Time Complexity                              :  O(N) where N is the number of nodes/size of the string to deserialize
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if(!root) return "";
        
        queue<TreeNode*> q;
        q.push(root);
        string ser = "";
        while(!q.empty()) {
            auto node = q.front();q.pop();
            if(node == nullptr) {
                ser += "null,";
                continue;
            }
            
            ser += to_string(node->val) + ',';
            q.push(node->left);
            q.push(node->right);
        }
        return ser;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if(data == "") return nullptr;
        int n = data.size();
        string temp = "";
        vector<string> str = stringSplit(data);
    
        TreeNode *root = new TreeNode(stoi(str[0]));
        queue<TreeNode*> q;
        q.push(root);
        int i=1;
        int sz = str.size();

        while(i<sz) {
            TreeNode *curr = q.front();q.pop();
            
            if(str[i] != "null") {
                curr->left = new TreeNode(stoi(str[i]));
                q.push(curr->left);
            }
            i++;
            
            if(i < sz and str[i] != "null") {
                curr->right = new TreeNode(stoi(str[i]));
                q.push(curr->right);
            }
            i++;
        }
        
        return root;
    }
    
    vector<string> stringSplit(string data) {
        vector<string> ans;
        string temp = "";
        for(int i=0;i<data.size();i++) {
            if(data[i] != ',') {
                temp += data[i]; 
            } else {
                ans.push_back(temp);
                temp = "";
            }
        }
        return ans;
    }
    
    
    
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));